### PR TITLE
[Diagnostics] Special case assignment between nominal types with optional promotion

### PR DESCRIPTION
**Pull request by xedin**
Skip fixing situation where source and destination of assignment are both
nominal types with different optionality until restriction is attempted.

Otherwise fix could be too greedy and diagnose valid code if all of the
types are known in advance.

Resolves: SR-13951
Resolves: rdar://problem/72166791

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/Sema/CSSimplify.cpp [hunk 1]
```diff
@@ -3730,6 +3730,7 @@ bool ConstraintSystem::repairFailures(
         // 2. "existential" check, which is handled by a missing conformance
         //    fix.
         if ((lhs->is<BoundGenericType>() && rhs->is<BoundGenericType>()) ||
+            (lhs->is<NominalType>() && rhs->is<NominalType>()) ||
```
Comment (LucianoPAlmeida) [inline lib/Sema/CSSimplify.cpp:3733]: Ah, I see :) 
Should we update the comment above to metion the new condition?

Comment (xedin) [reply lib/Sema/CSSimplify.cpp:3733]: I guess, I could also mention nominal there, it's going to be handled by `deep equality` just like generic types...

Comment (xedin) [reply lib/Sema/CSSimplify.cpp:3733]: Adjusted!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Review by LucianoPAlmeida — APPROVED** [review]

**Review by hborla — APPROVED** [review]
LGTM!

**Pull request merged**